File: Algorithm/Python/25/0003_Longest_Substring_Without_Repreating_Characters.py
'''

Given a string, find the length of the longest substring without repeating characters.

Example 1:

Input: "abcabcbb"
Output: 3 
Explanation: The answer is "abc", with the length of 3. 

'''

# use dictionary to record visited
# use enumerate(s) to speed up

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res = 0
        mp = {}
        i = 0

        for j in range(len(s)):
            if s[j] in mp:
                i = max(mp[s[j]] + 1, i)

            res = max(res, j - i + 1)
            mp[s[j]] = j

        return res

File: Algorithm/Java/0003_Longest_Substring_Without_Repreating_Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length(), res = 0;
        Map<Character, Integer> map = new HashMap<>();

        for (int i = 0, j = 0; j < n; j++) {
            if (map.containsKey(s.charAt(j))) {
                i = Math.max(map.get(s.charAt(j)) + 1, i);
            }

            res = Math.max(res, j - i + 1);
            map.put(s.charAt(j), j);
        }
        return res;
    }
}